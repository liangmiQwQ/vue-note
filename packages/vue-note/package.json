{
  "name": "vue-note",
  "type": "module",
  "version": "0.0.0-alpha.0",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/liangmiqwq/vue-note.git"
  },
  "bugs": "https://github.com/liangmiqwq/vue-note/issues",
  "keywords": ["Vuejs"],
  "sideEffects": false
}
