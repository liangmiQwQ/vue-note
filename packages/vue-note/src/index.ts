import type { Component } from 'vue'

export default {}

export declare function defineCommentComponents(component: () => any): Component
