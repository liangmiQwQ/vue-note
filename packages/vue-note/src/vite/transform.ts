import type { Rollup, TransformResult, ViteDevServer } from 'vite'
import type { FileParseResult } from '../compilor/parse'
import type { VueNoteQuery } from './query'
import { createHash } from 'node:crypto'
import { parseComponents } from '../compilor/component'
import { parse } from '../compilor/parse'
import { resolve } from '../compilor/resolve'
import { getUniqueID } from '../compilor/utils/id'

export interface TransformOption {
  server: ViteDevServer
  isProduction: boolean
}

export interface TransformHashCache {
  ast: string
  template: Map<string, string>
}

export async function transform(
  src: string,
  filename: string,
  ctx: Rollup.TransformPluginContext,
  query: VueNoteQuery,
  ssr: boolean,
  opt: TransformOption,
  _cache?: TransformHashCache,
): Promise<{ result?: TransformResult, cache?: TransformHashCache }> {
  if (query.raw)
    return {}

  const hmr = !ssr && opt.server && opt.server.config.server.hmr !== false && !opt.isProduction

  const fileParseResult = parse(src, filename, ctx) // get AST & raw components (scripts and templates)
  const cache = getCache(filename, fileParseResult)
  const compiledComponents = parseComponents(filename, fileParseResult.rawComponents)
  const resolvedCode = resolve(fileParseResult.astRestult.program, compiledComponents, ctx, hmr)

  let result: TransformResult
  const { rolldownVersion, transformWithOxc } = await import('vite')
  if (rolldownVersion) {
    const { code } = await transformWithOxc(
      resolvedCode,
      filename,
      {
        lang: 'ts',
      },
    )
    result = {
      code,
      map: null,
    }
  }
  else {
    const { transformWithEsbuild } = await import('vite')
    const { code } = await transformWithEsbuild(
      resolvedCode,
      filename,
      {
        target: 'esnext',
        charset: 'utf8',
        loader: 'ts',
      },
    )
    result = {
      code,
      map: null,
    }
  }

  return {
    result,
    cache,
  }
}

function getCache(filename: string, result: FileParseResult): TransformHashCache {
  const template = new Map<string, string>()

  result.rawComponents.forEach((e) => {
    template.set(getUniqueID(filename, e.id), createHash('md5').update(e.template).digest('hex'))
  })

  return {
    ast: createHash('md5').update(JSON.stringify(result.astRestult.program)).digest('hex'),
    template,
  }
}
