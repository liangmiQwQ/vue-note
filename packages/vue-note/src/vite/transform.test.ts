import type { Rollup } from 'vite'
import type { TransformOption } from './transform'
import { describe, expect, it, vi } from 'vitest'
import { transform } from './transform'

function createMockContext(): Rollup.TransformPluginContext {
  return {
    error: vi.fn(),
    warn: vi.fn(),
    info: vi.fn(),
    debug: vi.fn(),
    getCombinedSourcemap: vi.fn(),
  } as any
}

function createMockServer(isProduction = false, hmr = false): TransformOption {
  return {
    isProduction,
    server: { config: { server: { hmr } } },
  } as any
}

const testScript = `
import { ref } from 'vue'
import { defineCommentComponent, defineTemplate } from 'vue-note'

export default defineCommentComponent(() => {
  const message = ref('Good Morning')

  defineTemplate(/* @template
    <div> {{ message }} </div>
    <div> Hello World </div>
  */)
})
`

describe('production', () => {
  it('should work in production', async () => {
    expect(await transform(testScript, `test.ts`, createMockContext(), {}, false, createMockServer(true, false))).toMatchSnapshot()
  })

  it('should work with style imports', async () => {
    expect(await transform(`import 'style.css'; \n ${testScript}`, `test.ts`, createMockContext(), {}, false, createMockServer(true, false))).toMatchSnapshot()
  })
})

describe('development', () => {
  it('should work and inject HMR code', async () => {
    expect(await transform(testScript, `test.ts`, createMockContext(), {}, false, createMockServer(false, true))).toMatchSnapshot()
  })

  it('should detect script change and inject HMR code', async () => {
    const first = await transform(testScript, `test.ts`, createMockContext(), {}, false, createMockServer(false, true))

    const second = await transform(testScript.replace('Good Morning', 'Good Afternoon'), `test.ts`, createMockContext(), {}, false, createMockServer(false, true), first.cache)
    expect(second.result?.code).toContain('const __scriptChanged = true;')
    expect(second.result?.code).not.toContain('changed: true,')
  })

  it('should detect template change and inject HMR code', async () => {
    const first = await transform(testScript, `test.ts`, createMockContext(), {}, false, createMockServer(false, true))

    const second = await transform(testScript.replace('Hello World', 'Hello Vue Note'), `test.ts`, createMockContext(), {}, false, createMockServer(false, true), first.cache)
    expect(second.result?.code).not.toContain('const __scriptChanged = true;')
    expect(second.result?.code).toContain('changed: true,')
  })
})
