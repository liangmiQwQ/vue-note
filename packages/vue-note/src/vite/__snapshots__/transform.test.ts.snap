// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`development > should work and inject HMR code 1`] = `
{
  "cache": {
    "ast": "da50a3ae9b19c971f6be6113f209f39a",
    "template": Map {
      "8ae9ae3e" => "95d6a8b04b8e4d3b8ea3e00064f89f36",
    },
  },
  "result": {
    "code": "import { ref } from "vue";
import { defineComponent as _defineComponent } from "vue";
import { toDisplayString as _toDisplayString, createElementVNode as _createElementVNode, Fragment as _Fragment, openBlock as _openBlock, createElementBlock as _createElementBlock } from "vue";
export const __VUE_HMR_RENDER_FUNCTIONS__ = { "8ae9ae3e": {
	changed: false,
	render: function render(_ctx, _cache, $props, $setup, $data, $options) {
		return _openBlock(), _createElementBlock(_Fragment, null, [_createElementVNode("div", null, _toDisplayString($setup.message), 1), _cache[0] || (_cache[0] = _createElementVNode("div", null, " Hello World ", -1))], 64);
	}
} };
export default (() => {
	const _component = _defineComponent({ setup(__props, { expose: __expose }) {
		__expose();
		const message = ref("Good Morning");
		const __returned__ = { message };
		Object.defineProperty(__returned__, "__isScriptSetup", {
			enumerable: false,
			value: true
		});
		return __returned__;
	} });
	_component.render = __VUE_HMR_RENDER_FUNCTIONS__["8ae9ae3e"].render;
	_component.__hmrId = "8ae9ae3e";
	typeof __VUE_HMR_RUNTIME__ !== "undefined" && __VUE_HMR_RUNTIME__.createRecord(_component.__hmrId, _component);
	return _component;
})();
export const __scriptChanged = false;
if (import.meta.hot) {
	import.meta.hot.accept(async (mod) => {
		if (!mod) return;
		if (mod.__scriptChanged) {
			return window.location.reload();
		}
		Object.keys(mod.__VUE_HMR_RENDER_FUNCTIONS__).forEach((id) => {
			const component = mod.__VUE_HMR_RENDER_FUNCTIONS__[id];
			if (component && component.changed) {
				__VUE_HMR_RUNTIME__.rerender(id, component.render);
			}
		});
	});
}
",
    "map": null,
  },
}
`;

exports[`production > should work in production 1`] = `
{
  "cache": {
    "ast": "da50a3ae9b19c971f6be6113f209f39a",
    "template": Map {
      "8ae9ae3e" => "95d6a8b04b8e4d3b8ea3e00064f89f36",
    },
  },
  "result": {
    "code": "import { ref } from "vue";
import { defineComponent as _defineComponent } from "vue";
import { unref as _unref, toDisplayString as _toDisplayString, createElementVNode as _createElementVNode, Fragment as _Fragment, openBlock as _openBlock, createElementBlock as _createElementBlock } from "vue";
export default (() => {
	const _component = _defineComponent({ setup(__props) {
		const message = ref("Good Morning");
		return (_ctx, _cache) => {
			return _openBlock(), _createElementBlock(_Fragment, null, [_createElementVNode("div", null, _toDisplayString(_unref(message)), 1), _cache[0] || (_cache[0] = _createElementVNode("div", null, " Hello World ", -1))], 64);
		};
	} });
	return _component;
})();
",
    "map": null,
  },
}
`;

exports[`production > should work with style imports 1`] = `
{
  "cache": {
    "ast": "643891df79c91e6f1fc896851540d168",
    "template": Map {
      "8ae9ae3e" => "95d6a8b04b8e4d3b8ea3e00064f89f36",
    },
  },
  "result": {
    "code": "import "style.css";
import { ref } from "vue";
import { defineComponent as _defineComponent } from "vue";
import { unref as _unref, toDisplayString as _toDisplayString, createElementVNode as _createElementVNode, Fragment as _Fragment, openBlock as _openBlock, createElementBlock as _createElementBlock } from "vue";
export default (() => {
	const _component = _defineComponent({ setup(__props) {
		const message = ref("Good Morning");
		return (_ctx, _cache) => {
			return _openBlock(), _createElementBlock(_Fragment, null, [_createElementVNode("div", null, _toDisplayString(_unref(message)), 1), _cache[0] || (_cache[0] = _createElementVNode("div", null, " Hello World ", -1))], 64);
		};
	} });
	return _component;
})();
",
    "map": null,
  },
}
`;
